### PR TITLE
Creating application security checklist

### PR DESCRIPTION
**Pull request by AnshumanTripathi**
Creating an application security checklist.

### Issue
https://github.com/kubernetes/sig-security/issues/111

### Deploy Preview
https://deploy-preview-46326--kubernetes-io-main-staging.netlify.app/docs/concepts/security/application-security-checklist/

### CODE REVIEW

File: content/en/docs/concepts/security/application-security-checklist.md [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+Configure {{< glossary_tooltip text="Security Context" term_id="Security-Context" >}} for the pod-container.
+
+### Pod SecurityContext
+- [ ] Set `runAsNonRoot: true`.
+- [ ] Configure User permissions and access controls using `runAsUser` and `runAsGroup`.
+- [ ] Optionally add a supplementary group with `fsGroup` to access persistent volumes.
+- [ ] Appropriate Pod Security Standard policies are enforced on all application namespace pods.
+
+### Container Security Context
+- [ ] Disable privilege escalations using `allowPrivilegeEscalation: true`.
```
Comment (raesene) [inline content/en/docs/concepts/security/application-security-checklist.md:44]:
```suggestion
- [ ] Disable privilege escalations using `allowPrivilegeEscalation: false`.
```

File: content/en/docs/concepts/security/application-security-checklist.md [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+- [ ] Configure the root filesystem to be read-only with `readOnlyRootFilesystem: true`.
+- [ ] Avoid running privileged containers with `privileged: false`.
+- [ ] [Set appropriate Seccomp profiles.](/docs/tasks/configure-pod-container/security-context/#set-the-seccomp-profile-for-a-container)
+- [ ] [Configure appropriate AppArmor policies.](/docs/tutorials/security/apparmor/)
+- [ ] [Assign appropriate SELinux labels to the containers.](/docs/tasks/configure-pod-container/security-context/#assign-selinux-labels-to-a-container)
+
```
Comment (raesene) [inline content/en/docs/concepts/security/application-security-checklist.md:50]:
```suggestion
- [ ] Drop all capabilities from the containers, and add back only specific ones that are needed for operation of the container
```

File: content/en/docs/concepts/security/application-security-checklist.md [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+- [ ] Optionally add a supplimentary group with `fsGroup` to access persistent volumes.
+- [ ] Appropriate Pod Security Standard policies are enforced to all pods in the application namespace.
+
+### Container Security Context
+- [ ] Disable privilege escalations using `allowPrivilegeEscalation: false`.
+- [ ] Configure the root filesystem to be read-only with `readOnlyRootFilesystem: true`.
+- [ ] Avoid running privileged containers with `privileged: false`.
+- [ ] [Set appropriate Seccomp profiles.](/docs/tasks/configure-pod-container/security-context/#set-the-seccomp-profile-for-a-container)
+- [ ] [Configure appropriate AppArmor policies.](/docs/tutorials/security/apparmor/)
+- [ ] [Assign appropriate SELinux labels to the contianers.](/docs/tasks/configure-pod-container/security-context/#assign-selinux-labels-to-a-container)
```
Comment (mtardy) [inline content/en/docs/concepts/security/application-security-checklist.md:50]: little typo
```suggestion
- [ ] [Assign appropriate SELinux labels to the containers.](/docs/tasks/configure-pod-container/security-context/#assign-selinux-labels-to-a-container)
```

File: content/en/docs/concepts/security/application-security-checklist.md [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+- [ ] Drop all capabilities from the containers, and add back only specific ones that are needed for operation of the container.
+
+## Role Based Access Control (RBAC)
+- [ ] Permissions such as `create`, `patch`, `update` and `delete` should be only granted if necessary.
+- [ ] Avoid creating RBAC permissions to create or update roles which can lead to [privilege escalation](/docs/reference/access-authn-authz/rbac/#privilege-escalation-prevention-and-bootstrapping).
+
+The `create`, `update` and `delete` are generally agreed upon to be used judicously. The `patch` permission can [allow users to update labels on the namespace or deployments](/docs/concepts/security/rbac-good-practices/#namespace-modification) making them more vulnarable.
```
Comment (mtardy) [inline content/en/docs/concepts/security/application-security-checklist.md:57]: Another typo here, not a fan either of "making them more vulnerable", maybe something like "allowing them to escalate privilege" or something less stronger but more conveyor of the "attack" sense? Not sure.
```suggestion
The `create`, `update` and `delete` are generally agreed upon to be used judicously. The `patch` permission can [allow users to update labels on the namespace or deployments](/docs/concepts/security/rbac-good-practices/#namespace-modification) making them more vulnerable.
```

File: content/en/docs/concepts/security/application-security-checklist.md [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+
+## Runtime Classes
+- [ ] Configure appropriate runtime classes for containers.
+
+Some containers may require a different isolation level from what is provided by the the default runtime of the cluster. `runtimeClassname` can be used in a podspec to define a differnt runtime class.
```
Comment (mtardy) [inline content/en/docs/concepts/security/application-security-checklist.md:62]: Maybe we can develop here without linking to existing solutions (or maybe we can for CNCF projects?). Explaining that you can use virtualization, or kernel emulation like gvisor, for very sensitive workloads where you need stronger boundaries.

**Comment by mtardy** [comment]
This looks good :)! I think it's a very valuable checklist for app developers.

File: content/en/docs/concepts/security/application-security-checklist.md (hunk 5 above)
Comment (AnshumanTripathi) [reply content/en/docs/concepts/security/application-security-checklist.md:66]: Added a mention about using gVisor, kata-containers and confidential virtual machines. 👍

**Review by NissesSenap — COMMENTED** [review]
First of all, great job, I think this document is a great addition. 

But here are my thoughts.
First, my question is, for whom is this document written for? Is it the 85-90% of app developers that run an http API or an event based app in a non-critical environment that just wants a good base layer? Or are we trying to point to all security features that exist from an app developer point of view?

Personally, I think we should aim this document mainly towards the 85-90% of the users. 
Especially since there is a tendency to want to be able to complete everything on a checklist, but if it becomes too painful, you end up doing nothing.
See keep it relatively simple and provide some kind of leveling system.

I would add a header called ## base security. In there I would add the most basic stuff.

Then I would add another header, ## advanced security, or something like that.

I would keep everything you have written, but I would move these down to advanced security

- rbac, if you are using rbac in your http api, you are most likely doing something wrong. I would emphasize that really hard in the document. 
- runtime class, custom runetime classes isn't something that the majority of Kubenetes environment will use.
- Seccomp profiles, It's more or less impossible to setup unless you have some kind of tooling + since version 1.25 the default profile is enabled by default.
- SElinux, most developers don't know what it is and just like secomp profile you more or less need tooling to create something decent.
- AppArmor, even fewer people know what it is, and you need custom tooling to create something more advanced than default.

**Comment by AnshumanTripathi** [comment]
> First of all, great job, I think this document is a great addition.
> 
> But here are my thoughts. First, my question is, for whom is this document written for? Is it the 85-90% of app developers that run an http API or an event based app in a non-critical environment that just wants a good base layer? Or are we trying to point to all security features that exist from an app developer point of view?
> 
> Personally, I think we should aim this document mainly towards the 85-90% of the users. Especially since there is a tendency to want to be able to complete everything on a checklist, but if it becomes too painful, you end up doing nothing. See keep it relatively simple and provide some kind of leveling system.
> 
> I would add a header called ## base security. In there I would add the most basic stuff.
> 
> Then I would add another header, ## advanced security, or something like that.
> 
> I would keep everything you have written, but I would move these down to advanced security
> 
> * rbac, if you are using rbac in your http api, you are most likely doing something wrong. I would emphasize that really hard in the document.
> * runtime class, custom runetime classes isn't something that the majority of Kubenetes environment will use.
> * Seccomp profiles, It's more or less impossible to setup unless you have some kind of tooling + since version 1.25 the default profile is enabled by default.
> * SElinux, most developers don't know what it is and just like secomp profile you more or less need tooling to create something decent.
> * AppArmor, even fewer people know what it is, and you need custom tooling to create something more advanced than default.

Thanks for the detailed response. I like this format and I will try to update the document accordingly.
Regarding the target audience I did add this point

> This checklist assumes a `developer` is a Kubernetes cluster user that interacts with namespaced scope objects.
https://github.com/kubernetes/website/pull/46326/files#diff-5285725b7ae5cc6b415f12dbb2acf91aea311bf5dd5dc71851b29c216afcb9adR20

I did realize that identifying the target audience is key here and then keeping the list generic so that it can be used by everyone. Because app developers can consist of tools that be running cluster wide like log and metric collectors or policy agents. This is where the line between admins and developers start to blur. I'll try to add more details about the target audience for this list.

**Comment by AnshumanTripathi** [comment]
@NissesSenap I have updated the checklist structure based on your recommendation. 

Can you expand on 

> rbac, if you are using rbac in your http api, you are most likely doing something wrong. I would emphasize that really hard in the document.

I'm not sure what you mean here.

**Comment by NissesSenap** [comment]
> @NissesSenap I have updated the checklist structure based on your recommendation. 
> 
> Can you expand on 
> 
> > rbac, if you are using rbac in your http api, you are most likely doing something wrong. I would emphasize that really hard in the document.
> 
> I'm not sure what you mean here.

That is great @AnshumanTripathi , I will try to take a deeper look at the update as soon as possible. My point is that the majority of all apps ruining in Kubernetes don't need access to its API. So I would emphasize that custom rbac rules for apps just aren't needed, in the majority of applications. Still good to talk about it, just like you have, but point to that there are limited use cases for it.

**Comment by AnshumanTripathi** [comment]
> > @NissesSenap I have updated the checklist structure based on your recommendation.
> > Can you expand on
> > > rbac, if you are using rbac in your http api, you are most likely doing something wrong. I would emphasize that really hard in the document.
> > 
> > 
> > I'm not sure what you mean here.
> 
> That is great @AnshumanTripathi , I will try to take a deeper look at the update as soon as possible. My point is that the majority of all apps ruining in Kubernetes don't need access to its API. So I would emphasize that custom rbac rules for apps just aren't needed, in the majority of applications. Still good to talk about it, just like you have, but point to that there are limited use cases for it.


Makes sense. I am not covering applications using Kubernetes client as a part of this checklist. I think having a doc about secure practices for applications that use the Kubernetes API (and operator) can be very helpful. I will create an issue for this :)

**Comment by AnshumanTripathi** [comment]
@NissesSenap created https://github.com/kubernetes/sig-security/issues/121 for hardening guide for using Kubernetes API

**Comment by tengqm** [comment]
I'd suggest we place this under `content/en/docs/reference/security/`, a new directory for consolidating security related contents. This page is not suitable for the concepts section.

**Comment by AnshumanTripathi** [comment]
> I'd suggest we place this under `content/en/docs/reference/security/`, a new directory for consolidating security related contents. This page is not suitable for the concepts section.

Actually in that case we can move the hardening guides to `content/en/docs/reference/issues-security` which already exists. It makes sense. I can move the [existing hardening guide](https://github.com/kubernetes/website/blob/main/content/en/docs/concepts/security/hardening-guide/authentication-mechanisms.md) an this guide under that section. But I dont think I should do that as a part of this change though?

**Comment by AnshumanTripathi** [comment]
> I'd suggest we place this under `content/en/docs/reference/security/`, a new directory for consolidating security related contents. This page is not suitable for the concepts section.

Actually we can move this checklist and the [existing checklist](https://github.com/kubernetes/website/blob/main/content/en/docs/concepts/security/security-checklist.md) to `content/en/docs/reference/issues-security` maybe? I dont think I should do that as a part of this change though.

**Comment by tengqm** [comment]
> > I'd suggest we place this under `content/en/docs/reference/security/`, a new directory for consolidating security related contents. This page is not suitable for the concepts section.
> 
> Actually we can move this checklist and the [existing checklist](https://github.com/kubernetes/website/blob/main/content/en/docs/concepts/security/security-checklist.md) to `content/en/docs/reference/issues-security` maybe? I dont think I should do that as a part of this change though.

The `issues-security` directory is positioned to be a folder for disclosing the vulnerabilities of kubernetes release. What we are trying to compile is a collection of guidelines, best practices for users.

I don't think we want to kick this in and then follow up with another PR to move it.
We may want to get the page into a suitable folder in the first place.
The migration of other relevant pages could be left as follow-ups, for sure.

**Comment by AnshumanTripathi** [comment]
> > > I'd suggest we place this under `content/en/docs/reference/security/`, a new directory for consolidating security related contents. This page is not suitable for the concepts section.
> > 
> > 
> > Actually we can move this checklist and the [existing checklist](https://github.com/kubernetes/website/blob/main/content/en/docs/concepts/security/security-checklist.md) to `content/en/docs/reference/issues-security` maybe? I dont think I should do that as a part of this change though.
> 
> The `issues-security` directory is positioned to be a folder for disclosing the vulnerabilities of kubernetes release. What we are trying to compile is a collection of guidelines, best practices for users.
> 
> I don't think we want to kick this in and then follow up with another PR to move it. We may want to get the page into a suitable folder in the first place. The migration of other relevant pages could be left as follow-ups, for sure.

Makes sense. I'll try to make that change

File: content/en/docs/concepts/security/application-security-checklist.md [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+- [ ] [Assign appropriate SELinux labels to the contianers.](/docs/tasks/configure-pod-container/security-context/#assign-selinux-labels-to-a-container)
+
+### Runtime Classes
+- [ ] Configure appropriate runtime classes for containers.
+
+Some containers may require a different isolation level from what is provided by the default runtime of the cluster. `runtimeClassname` can be used in a podspec to define a differnt runtime class.
+
+For sensitive workloads consider using Kernel emulation tools like [gVisor](https://gvisor.dev/docs/) or [kata-containers](https://katacontainers.io/).
+
+In high trust environments, consider using [confidential virutal machines](/blog/2023/07/06/confidential-kubernetes/) to improve cluster security even further.
```
Comment (sftim) [inline content/en/docs/concepts/security/application-security-checklist.md:84]:
```suggestion
In high trust environments, consider using [confidential virtual machines](/blog/2023/07/06/confidential-kubernetes/) to improve cluster security even further.
```

Watch out for linking to non-evergreen blog articles (anything that gets marked as out of date once at least a year old).

File: content/en/docs/concepts/security/application-security-checklist.md [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+- [ ] [Set appropriate Seccomp profiles.](/docs/tasks/configure-pod-container/security-context/#set-the-seccomp-profile-for-a-container)
+- [ ] [Configure appropriate AppArmor policies.](/docs/tutorials/security/apparmor/)
+- [ ] [Assign appropriate SELinux labels to the contianers.](/docs/tasks/configure-pod-container/security-context/#assign-selinux-labels-to-a-container)
+
+### Runtime Classes
+- [ ] Configure appropriate runtime classes for containers.
+
+Some containers may require a different isolation level from what is provided by the default runtime of the cluster. `runtimeClassname` can be used in a podspec to define a differnt runtime class.
+
+For sensitive workloads consider using Kernel emulation tools like [gVisor](https://gvisor.dev/docs/) or [kata-containers](https://katacontainers.io/).
```
Comment (sftim) [inline content/en/docs/concepts/security/application-security-checklist.md:82]: Kata containers is not a kernel emulator:
```suggestion
For sensitive workloads consider using kernel emulation tools like [gVisor](https://gvisor.dev/docs/), or virtualized isolation using a mechanism such as [kata-containers](https://katacontainers.io/).
```

File: content/en/docs/concepts/security/application-security-checklist.md
Comment (sftim) [inline content/en/docs/concepts/security/application-security-checklist.md:1]: - This page needs a third party links disclaimer; see https://kubernetes.io/docs/contribute/style/hugo-shortcodes/#third-party-content-marker
- API verbs like **apply** and **create** are written as lower case bold. We use upper case non bold when documenting verbs at the underlying HTTP level (eg OPTIONS, PUT, GET).

- (nit) When a sentence ends with a period, ie `.`, don't include the period inside the anchor text of a hyperlink.

File: content/en/docs/concepts/security/application-security-checklist.md [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+Checklists are **not** sufficient for attaining a good security posture on their own. A good security posture requires constant attention and improvement, but a checklist can be the first step on the never-ending journey towards security preparedness. Some of the recommendations in this checklist may be too restrictive or too lax for your specific security needs. Since Kubernetes security is not "one size fits all", each category of checklist items should be evaluated on its merits.
+{{< /caution >}}
+
+
+<!-- body -->
+## Base security hardening
+The following checklist provide a base security hardening recommendations that would apply to most applications deploying to Kubernetes.
+
+### Application design
+- [ ] Following the right [design principles for the application.](https://kubernetes.io/blog/2018/03/principles-of-container-app-design/)
```
Comment (sftim) [inline content/en/docs/concepts/security/application-security-checklist.md:33]: I don't recommend linking to that article. Find a better source, such as https://www.cncf.io/wp-content/uploads/2022/06/CNCF_cloud-native-security-whitepaper-May2022-v2.pdf

File: content/en/docs/concepts/security/application-security-checklist.md [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+- [ ] Application configured with appropriate {{< glossary_tooltip text="QoS class" term_id="QoS-class" >}} through resource request and limits
+  - [ ] Memory limit is set for the workloads with a limit equal or inferior to the request.
+  - [ ] CPU limit might be set on sensitive workloads.
+
+### Service account
+- [ ] Avoid using `default` service account. Create service accounts for workloads.
```
Comment (sftim) [inline content/en/docs/concepts/security/application-security-checklist.md:39]:
```suggestion
- [ ] Avoid using the `default` ServiceAccount. Create ServiceAccounts for each workload or microservice.
```

File: content/en/docs/concepts/security/application-security-checklist.md [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+- [ ] Optionally add a supplimentary group with `fsGroup` to access persistent volumes.
+- [ ] Appropriate Pod Security Standard policies are enforced to all pods in the application namespace.
+
+### Container Security Context
+- [ ] Disable privilege escalations using `allowPrivilegeEscalation: false`.
+- [ ] Configure the root filesystem to be read-only with `readOnlyRootFilesystem: true`.
+- [ ] Avoid running privileged containers with `privileged: false`.
+- [ ] Drop all capabilities from the containers, and add back only specific ones that are needed for operation of the container.
+
+### Role Based Access Control (RBAC)
```
Comment (sftim) [inline content/en/docs/concepts/security/application-security-checklist.md:54]: (nit)
```suggestion
### Role Based Access Control (RBAC) {#rbac}

These suggestions apply if your cluster uses
Kubernetes [RBAC](https://kubernetes.io/docs/reference/access-authn-authz/rbac/)
for authorization, or for applications that need to work across different kinds of cluster.

```

File: content/en/docs/concepts/security/application-security-checklist.md [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+- [ ] Review bindings for the `system:unauthenticated` group and remove them where possible, as this gives access to anyone who can contact the API server at a network level.
+
+The `create`, `update` and `delete` should be used judicously. The `patch` permission can [allow users to update labels on the namespace or deployments](/docs/concepts/security/rbac-good-practices/#namespace-modification) which can increase the attack surface.
+
+### Image security
+- [ ] Using a image scanning tool the scan an image before deploying containers in the Kubernetes cluster.
+- [ ] Use container signing to validate the container image signature before deploying to the Kubernetes cluster.
+
+### Network Policies
+- [ ] Configure network policies to only allow expected ingress and egress traffic from the pods.
```
Comment (sftim) [inline content/en/docs/concepts/security/application-security-checklist.md:66]: How about:
```suggestion
- [ ] Configure [NetworkPolicies](/docs/concepts/services-networking/network-policies/)
      to only allow expected ingress and egress traffic from the pods.
      Make sure that your cluster provides and enforces NetworkPolicy. If you are writing an application
      that people will deploy to different clusters, consider whether you can assume that NetworkPolicy
      is available and enforced.
```

File: content/en/docs/concepts/security/application-security-checklist.md [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+
+## Advanced Security Hardening
```
Comment (sftim) [inline content/en/docs/concepts/security/application-security-checklist.md:68]:
```suggestion
## Advanced security hardening {#advanced}

```

File: content/en/docs/concepts/security/application-security-checklist.md [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+This section of this guide covers some advanced security hardening points which might be valuable based on different Kubernetes environment setup
+
+### Linux container security
+Configure {{< glossary_tooltip text="Security Context" term_id="Security-Context" >}} for the pod-container.
```
Comment (sftim) [inline content/en/docs/concepts/security/application-security-checklist.md:72]:
```suggestion

Configure a {{< glossary_tooltip text="securityContext" term_id="Security-Context" >}} for the pod and / or container.
```

File: content/en/docs/concepts/security/application-security-checklist.md [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+- [ ] [Set appropriate Seccomp profiles.](/docs/tasks/configure-pod-container/security-context/#set-the-seccomp-profile-for-a-container)
+- [ ] [Configure appropriate AppArmor policies.](/docs/tutorials/security/apparmor/)
+- [ ] [Assign appropriate SELinux labels to the contianers.](/docs/tasks/configure-pod-container/security-context/#assign-selinux-labels-to-a-container)
```
Comment (sftim) [inline content/en/docs/concepts/security/application-security-checklist.md:75]:
```suggestion
- [ ] [Assign appropriate SELinux labels to the containers.](/docs/tasks/configure-pod-container/security-context/#assign-selinux-labels-to-a-container)
```

File: content/en/docs/concepts/security/application-security-checklist.md [hunk 15]
```diff
@@ -0,0 +1,84 @@
+---
+title: Application Security Checklist
+description: >
+  Baseline checklist for ensuring application security for Kubernetes Developers
```
Comment (sftim) [inline content/en/docs/concepts/security/application-security-checklist.md:4]:
```suggestion
  Baseline guidelines around ensuring application security on Kubernetes, aimed at application developers
```

File: content/en/docs/concepts/security/application-security-checklist.md [hunk 16]
```diff
@@ -0,0 +1,84 @@
...
+  - [ ] CPU limit might be set on sensitive workloads.
+
+### Service account
+- [ ] Avoid using `default` service account. Create service accounts for workloads.
+- [ ] `automountServiceAccountToken` should be set to `false` unless the pod specifically requires access to the Kubernetes API to operate.
+
+### Pod Security Context
+- [ ] Set `runAsNonRoot: true`.
+- [ ] Configure User permissions and access controls using `runAsUser` and `runAsGroup`.
+- [ ] Optionally add a supplimentary group with `fsGroup` to access persistent volumes.
```
Comment (sftim) [inline content/en/docs/concepts/security/application-security-checklist.md:45]:
```suggestion
- [ ] Optionally add a supplementary group with `fsGroup` to access persistent volumes.
```

File: content/en/docs/concepts/security/application-security-checklist.md [hunk 17]
```diff
@@ -0,0 +1,84 @@
...
+  - [ ] Memory limit is set for the workloads with a limit equal or inferior to the request.
+  - [ ] CPU limit might be set on sensitive workloads.
+
+### Service account
+- [ ] Avoid using `default` service account. Create service accounts for workloads.
+- [ ] `automountServiceAccountToken` should be set to `false` unless the pod specifically requires access to the Kubernetes API to operate.
+
+### Pod Security Context
+- [ ] Set `runAsNonRoot: true`.
+- [ ] Configure User permissions and access controls using `runAsUser` and `runAsGroup`.
```
Comment (sftim) [inline content/en/docs/concepts/security/application-security-checklist.md:44]:
```suggestion
- [ ] Configure the container to execute as a less privileged user (for example, using
      `runAsUser` and `runAsGroup`), and configure appropriate permissions on files or directories
      inside the container image.
```

File: content/en/docs/concepts/security/application-security-checklist.md [hunk 18]
```diff
@@ -0,0 +1,84 @@
...
+- [ ] Optionally add a supplimentary group with `fsGroup` to access persistent volumes.
+- [ ] Appropriate Pod Security Standard policies are enforced to all pods in the application namespace.
+
+### Container Security Context
+- [ ] Disable privilege escalations using `allowPrivilegeEscalation: false`.
+- [ ] Configure the root filesystem to be read-only with `readOnlyRootFilesystem: true`.
+- [ ] Avoid running privileged containers with `privileged: false`.
```
Comment (sftim) [inline content/en/docs/concepts/security/application-security-checklist.md:51]:
```suggestion
- [ ] Avoid running privileged containers (set `privileged: false`).
```

File: content/en/docs/concepts/security/application-security-checklist.md [hunk 19]
```diff
@@ -0,0 +1,84 @@
...
+- [ ] Following the right [design principles for the application.](https://kubernetes.io/blog/2018/03/principles-of-container-app-design/)
+- [ ] Application configured with appropriate {{< glossary_tooltip text="QoS class" term_id="QoS-class" >}} through resource request and limits
+  - [ ] Memory limit is set for the workloads with a limit equal or inferior to the request.
+  - [ ] CPU limit might be set on sensitive workloads.
+
+### Service account
+- [ ] Avoid using `default` service account. Create service accounts for workloads.
+- [ ] `automountServiceAccountToken` should be set to `false` unless the pod specifically requires access to the Kubernetes API to operate.
+
+### Pod Security Context
```
Comment (sftim) [inline content/en/docs/concepts/security/application-security-checklist.md:42]:
```suggestion
### Pod-level `securityContext` recommendations {#security-context-pod}
```

File: content/en/docs/concepts/security/application-security-checklist.md [hunk 20]
```diff
@@ -0,0 +1,84 @@
...
+- [ ] Set `runAsNonRoot: true`.
+- [ ] Configure User permissions and access controls using `runAsUser` and `runAsGroup`.
+- [ ] Optionally add a supplimentary group with `fsGroup` to access persistent volumes.
+- [ ] Appropriate Pod Security Standard policies are enforced to all pods in the application namespace.
+
+### Container Security Context
```
Comment (sftim) [inline content/en/docs/concepts/security/application-security-checklist.md:48]:
```suggestion
### Container-level `securityContext` recommendations {#security-context-container}

```

File: content/en/docs/concepts/security/application-security-checklist.md [hunk 21]
```diff
@@ -0,0 +1,84 @@
...
+
+### Service account
+- [ ] Avoid using `default` service account. Create service accounts for workloads.
+- [ ] `automountServiceAccountToken` should be set to `false` unless the pod specifically requires access to the Kubernetes API to operate.
+
+### Pod Security Context
+- [ ] Set `runAsNonRoot: true`.
+- [ ] Configure User permissions and access controls using `runAsUser` and `runAsGroup`.
+- [ ] Optionally add a supplimentary group with `fsGroup` to access persistent volumes.
+- [ ] Appropriate Pod Security Standard policies are enforced to all pods in the application namespace.
```
Comment (sftim) [inline content/en/docs/concepts/security/application-security-checklist.md:46]: Maybe this:
```suggestion
- [ ] The application deploys into a namespace that enforces an appropriate _Pod security standard_.
      If you cannot control this enforcement for the cluster(s) where the application is deployed,
      take this into account either through documentation or additional defense in depth.
```

File: content/en/docs/concepts/security/application-security-checklist.md [hunk 22]
```diff
@@ -0,0 +1,84 @@
...
+- [ ] Configure the root filesystem to be read-only with `readOnlyRootFilesystem: true`.
+- [ ] Avoid running privileged containers with `privileged: false`.
+- [ ] Drop all capabilities from the containers, and add back only specific ones that are needed for operation of the container.
+
+### Role Based Access Control (RBAC)
+- [ ] Permissions such as `create`, `patch`, `update` and `delete` should be only granted if necessary.
+- [ ] Avoid creating RBAC permissions to create or update roles which can lead to [privilege escalation](/docs/reference/access-authn-authz/rbac/#privilege-escalation-prevention-and-bootstrapping).
+- [ ] Review bindings for the `system:unauthenticated` group and remove them where possible, as this gives access to anyone who can contact the API server at a network level.
+
+The `create`, `update` and `delete` should be used judicously. The `patch` permission can [allow users to update labels on the namespace or deployments](/docs/concepts/security/rbac-good-practices/#namespace-modification) which can increase the attack surface.
```
Comment (sftim) [inline content/en/docs/concepts/security/application-security-checklist.md:59]: Try this:
```suggestion
The **create**, **update** and **delete** verbs should be permitted judiciously. The **patch** verb if allowed on a Namespace can
[allow users to update labels on the namespace or deployments](/docs/concepts/security/rbac-good-practices/#namespace-modification) which can increase the attack surface.

For sensitive workloads, consider providing a recommended ValidatingAdmissionPolicy that further restricts the permitted write actions.
```

**Review by sftim — COMMENTED** [review]
Thanks! We could really use this new page.

Here's some advice about how to get this ready to merge.

**Comment by tengqm** [comment]
This page, in its current format, with its current contents, doesn't belong to the concepts section. Please consider place it under the references section.

**Comment by reylejano** [comment]
I think this proposed "application security checklist" aligns with the live ["security checklist" page](https://kubernetes.io/docs/concepts/security/security-checklist/) which is aimed towards operators/cluster admins. 
Since the "security checklist" page is currently under concepts, I'm ok to merge the "application security checklist" since under concepts

If we decide references is a better place then a PR can be made to move both of the "checklists" and possibly more like the 
[Authentication Mechanism Hardening Guide](https://kubernetes.io/docs/concepts/security/hardening-guide/authentication-mechanisms/)

/approve

**Pull request merged**